src/fluree/db/ledger/indexing.clj

```diff
@@ -398,53 +400,134 @@
   [session]
-  (-> session :conn :meta :novelty-min))
+  (let [meta (-> session :conn :meta)]
+    [(:novelty-min meta) (:novelty-max meta)]))
 
 
-(defn index*
-  ([session {:keys [remove-preds] :as opts}]
-   (go-try
-     (if (session/indexing? session)
-       false
-       (let [latest-db     (<? (session/current-db session))
-             novelty-size  (get-in latest-db [:novelty :size])
-             novelty-min   (novelty-min session)
-             remove-preds? (and (not (nil? remove-preds)) (seq remove-preds))
-             needs-index?  (or remove-preds? (>= novelty-size novelty-min))]
-         (if needs-index?
-           ;; kick off indexing with this DB
-           (<? (index* session latest-db opts))
-           ;; no index needed, return false
-           false)))))
-  ([session db opts]
+(defn novelty-min?
+  "Returns true if ledger is beyond novelty-min threshold."
+  [session db]
+  (let [[min _] (novelty-min-max session)
+        novelty-size (get-in db [:novelty :size])]
+    (> novelty-size min)))
+
+
+(defn novelty-max?
+  "Returns true if ledger is beyond novelty-max threshold."
+  [session db]
+  (let [[_ max] (novelty-min-max session)
+        novelty-size (get-in db [:novelty :size])]
+    (> novelty-size max)))
+
+
+(defn reindexed-db
+  "If a db is being reindexed, this will return the db if it is complete, else nil."
+  [session]
+  (some-> (session/indexing-promise-ch session)
+          async/poll!))
+
+
+(defn do-index
+  "Performs an index operation and returns a promise-channel of the latest db once complete"
+  [session db remove-preds]
+  (let [[lock? pc] (session/acquire-indexing-lock! session (async/promise-chan))]
+    (when lock?
+      ;; when we have a lock, reindex and put updated db onto pc.
+      (async/go
+        (let [indexed-db (<? (index db {:remove-preds remove-preds}))]
+          (<? (txproto/write-index-point-async (-> db :conn :group) indexed-db))
+          ;; updated-db might have had additional block(s) written to it, so instead
+          ;; of using it, reload from disk.
+          (session/clear-db! session)                       ;; clear db cache to force reload
+          (async/put! pc indexed-db))))
+    pc))
+
+
+(defn merge-index-block
+  "Attempts to merge newly completed index job with block-map.
+  Optimistically the new index will be one block behind the block-map,
+  but if not we will pause momentarily to try to see if things catch up
+  before throwing and error."
+  [session indexed-db {:keys [block flakes] :as block-map}]
+  (go-try
+    (let [indexed-block (:block indexed-db)
+          current?      (= indexed-block block)]
+      (log/warn "-------- merge-index-block: " current? indexed-block)
+      (if current?
+        (do
+          (session/release-indexing-lock! session indexed-block)
+          (assoc block-map :db-after indexed-db))
+        (let [_         (log/warn "^^^^^^ PAUSING  ^^^^ merge-index-block: " indexed-block block)
```

Comment:
Is there a session ID of some sort we can output to correlate this with the previous log message even if other things get logged in between? Might not be much of an issue in practice.

Reply:
Fixed logging I left in by mistake.